### PR TITLE
pyflyte run command with first class workflow input parameters

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Updates the run command with first class workflow options and typed workflow input parameters.

This is accomplished by using a nested multi command so that workflows for a specific file can be listed in the help:

```
$ pyflyte run example.py --help
Usage: pyflyte run example.py [OPTIONS] COMMAND [ARGS]...

  Run a workflow in a file using script mode

Options:
  --help  Show this message and exit.

Commands:
  typechecker_dataclass   Run example.typechecker_dataclass in script mode
  typechecker_datetime    Run example.typechecker_datetime in script mode
  typechecker_json_types  Run example.typechecker_json_types in script mode
  typechecker_primitives  Run example.typechecker_primitives in script mode
  wf                      Run example.wf in script mode
```

And workflow inputs are subsequently listed:

```
$ pyflyte run example.py wf --help
Usage: pyflyte run example.py wf [OPTIONS]

  Run example.wf in script mode

Options:
  --remote                Whether to register and run the workflow on a Flyte
                          deployment
  -p, --project TEXT      Project to register and run this workflow in
  -d, --domain TEXT       Domain to register and run this workflow in
  --destination-dir TEXT  Directory inside the image where the tar file
                          containing the code will be copied to
  -i, --image TEXT        Image used to register and run.
  --service-account TEXT  Service account used when executing this workflow
  --wait-execution        Whether wait for the execution to finish
  --dump-snippet          Whether dump a code snippet instructing how to load
                          the workflow execution using flyteremote
  --n INTEGER             [default: 200]
  --mean FLOAT            [default: 0.0]
  --sigma FLOAT           [default: 1.0]
  --help                  Show this message and exit.
```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2241

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/clis/sdk_in_container/run.py [hunk 1]
```diff
@@ -30,115 +31,217 @@
...
+
+    def convert(self, value, param, ctx) -> typing.Optional[typing.Union[typing.Dict, typing.List]]:
+        if value is None:
+            return None
+        if isinstance(value, list) or isinstance(value, dict):
+            return value
+        return json.loads(value)
+
+
+class DataframeType(click.ParamType):
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/run.py:48]: for pandas dataframe we read the entire parquet. this is not good. It should just be transparently sent!

File: flytekit/clis/sdk_in_container/run.py [hunk 2]
```diff
@@ -30,115 +31,217 @@
...
+
+def set_is_remote(ctx: click.Context, param: str, value: str):
+    ctx.obj[REMOTE_KEY] = bool(value)
+
+
+def get_workflow_command_base_params() -> typing.List[click.Option]:
+    """
+    Return the set of base parameters added to every pyflyte run workflow subcommand.
+    """
+    return [
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/run.py:109]: lets also take an argument for `--name` - execution_name

File: flytekit/clis/sdk_in_container/run.py (hunk 1 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/run.py:48]: I think we should just create a dummy object called
@dataclass 
class DataframeType():
   file_loc: os.PathLike

File: flytekit/clis/sdk_in_container/run.py [hunk 3]
```diff
@@ -293,3 +367,11 @@ def encode(
             uri=df_remote_location.native_url[: -len(filename)],
             metadata=literals.StructuredDatasetMetadata(structured_dataset_type),
         )
+
+
+run = RunCommand(
+    name="run",
+    help="Run_old command, a.k.a. script mode. It allows for a a single script to be "
+    + "registered and run from the command line or any interactive environment "
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/run.py:375]:
```suggestion
    + "registered and run from the command line. "
```

This does not help to run it from jupyter, lets drop that

Comment (katrogan) [reply flytekit/clis/sdk_in_container/run.py]: done

File: flytekit/clis/sdk_in_container/run.py (hunk 1 above)
Comment (katrogan) [reply flytekit/clis/sdk_in_container/run.py:48]: I think this is okay, we only read it in the local case and even though the type engine will write the df to a file and read from it later, I don't think local datasets are going to large enough to really affect performance.

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — DISMISSED** [review]

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
